File: Problems/P10990-f.py
n, m = map(int, input().split())
tag = [[0] * 21 for _ in range(n + 1)]
tim = [[0] * 21 for _ in range(n + 1)]

for i in range(1, m + 1):
    ln = [int(i) for i in input().split()]
    op = ln[0]
    if op == 1:
        x, y, z = ln[1:]
        while x and y >= 0:
            tag[x][min(y, 20)] = z
            tim[x][min(y, 20)] = i
            x >>= 1; y -= 1
    else:
        x = ln[1]
        y = 0; t = 0; res = 0
        while x:
            for j in range(y, 21):
                if tim[x][j] > t:
                    t = tim[x][j]
                    res = tag[x][j]
            x >>= 1; y += 1
        print(res)

File: Problems/P10990-f.cpp
#include <bits/stdc++.h>

#define i64 long long
#define rep(i, l, r) for (int i = (l); i <= r; ++i)

constexpr int N = 1e6 + 5;

int n, m;
int tag[N][21];
int tim[N][21];

int main() {
    std::cin >> n >> m;
    rep (i, 1, m) {
        int op; std::cin >> op;
        if (op == 1) {
            int x, y, z; std::cin >> x >> y >> z;
            while (x && y >= 0) {
                tag[x][std::min(y, 20)] = z;
                tim[x][std::min(y, 20)] = i;
                x >>= 1; --y;
            }
        }
        else {
            int x = 0, y = 0; int t = 0, res = 0;
            std::cin >> x;
            while (x) {
                rep (j, y, 20) if (tim[x][j] > t)
                    t = tim[x][j], res = tag[x][j];
                x >>= 1; ++y;
            }
            std::cout << res << '\n';
        }
    }
}